packages/controls/src/controls/definition.ts

```diff
@@ -0,0 +1,101 @@
+import { z } from 'zod'
+
+import { type ParseResult } from '../lib/zod'
+
+import { type Data } from '../common/types'
+import {
+  type CopyContext,
+  type MergeContext,
+  type MergeTranslatableDataContext,
+} from '../context'
+import { type IntrospectionTarget } from '../introspection'
+import {
+  Serializable,
+  serializeObject,
+  type SerializedRecord,
+} from '../serialization'
+
+import { ControlInstance, type SendMessage } from './instance'
+
+export type SchemaType<T> = z.ZodType<T>
+export type SchemaTypeAny = SchemaType<any> | z.ZodBranded<SchemaType<any>, any>
+
+export abstract class ControlDefinition<
+  ControlType extends string = string,
+  Config = unknown,
+  DataType extends Data = Data,
+  ValueType extends Data = Data,
+  ResolvedValueType = Data | unknown,
+  InstanceType extends ControlInstance<any> = ControlInstance<any>,
+> extends Serializable {
+  constructor(readonly config: Config) {
+    super()
+  }
+
+  abstract get controlType(): ControlType
+
+  abstract get schema(): {
+    definition: SchemaType<unknown>
+    type: SchemaType<ControlType>
+    data: SchemaType<DataType>
+    value: SchemaType<ValueType>
+    resolvedValue: SchemaType<ResolvedValueType>
+  }
+
+  abstract safeParse(
+    data: unknown | undefined,
+  ): ParseResult<DataType | undefined>
+
+  abstract fromData(data: DataType | undefined): ValueType | undefined
+  abstract toData(value: ValueType): DataType
+
+  abstract copyData(
+    data: DataType | undefined,
+    context: CopyContext,
+  ): DataType | undefined
+
+  mergeData(
+    base: DataType,
+    override: DataType = base,
+    _context: MergeContext,
+  ): DataType {
+    return override
+  }
+
+  getTranslatableData(_data: DataType): Data {
+    return null
+  }
+
+  mergeTranslatedData(
+    data: DataType,
+    _translatedData: Data,
+    _context: MergeTranslatableDataContext,
+  ): Data {
+    return data as Data
+  }
+
+  abstract createInstance(sendMessage: SendMessage<any>): InstanceType
+
+  abstract serialize(): [SerializedRecord, Transferable[]]
+
+  introspect<R>(
+    data: DataType | undefined,
+    target: IntrospectionTarget<R>,
+  ): R[] {
+    return target.introspect(data)
+  }
+}
+
+export function serialize(
+  config: unknown,
+  rest: { type: string } & Record<string, unknown>,
+): [SerializedRecord, Transferable[]] {
```

Comment:
High-level serialization helper. Most controls call this to implement the `serialize` method.